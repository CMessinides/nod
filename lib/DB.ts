import { Pool } from "pg";
import { connectionString } from "../config/db.config";
import { dev } from "../config/server.config";

const pool = new Pool({ connectionString, ssl: !dev });

export default {
	query: pool.query
};
