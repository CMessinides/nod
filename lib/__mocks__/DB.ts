const DB = jest.genMockFromModule("../DB").default;

export function mockRows(rows = [], { once = true } = {}): void {
	const result = { rows, rowCount: rows.length };
	if (once) {
		DB.query.mockResolvedValueOnce(result);
	} else {
		DB.query.mockResolvedValue(result);
	}
}

DB.__mockRows = mockRows;
DB.__mockRows([], { once: false });

export default DB;
