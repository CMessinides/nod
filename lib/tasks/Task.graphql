type Task {
	id: ID!
	name: String!
	done: Boolean!
	createdAt: Date!
}
