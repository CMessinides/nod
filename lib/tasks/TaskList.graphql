type TaskList {
	id: ID!
	name: String!
	tasks: [Task]!
}
