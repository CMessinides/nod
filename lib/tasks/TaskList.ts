import { TaskListModel } from "./interfaces";
import DB from "../DB";
import Task from "./Task";

type TaskListRecord = {
	id: number;
	name: string;
};

type TaskListParams = Omit<TaskListModel, "tasks">;

export default class TaskList implements TaskListModel {
	id: number;
	name: string;

	static async findById(id: number) {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM task_lists WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return TaskList.fromRecord(rows[0] as TaskListRecord);
	}

	static fromRecord(record: TaskListRecord) {
		return new TaskList(record);
	}

	constructor(params: TaskListParams) {
		for (const key of Object.keys(params)) {
			this[key] = params[key];
		}
	}

	tasks() {
		return Task.getAllInList(this.id);
	}
}
