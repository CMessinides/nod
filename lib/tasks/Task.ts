import DB from "../DB";
import { TaskModel } from "./interfaces";
import TaskList from "./TaskList";
import LinkedList from "../LinkedList";

interface TaskRecord {
	id: number;
	name: string;
	done: boolean;
	created_at: Date;
	list_id: number;
	prev_task_id: number;
}

type TaskParams = Omit<TaskModel, "list"> & {
	listId: number;
};

export default class Task implements TaskModel {
	id: number;
	name: string;
	done: boolean;
	createdAt: Date;
	private listId: number;

	static async findById(id: number): Promise<Task> {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM tasks WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return Task.fromRecord(rows[0] as TaskRecord);
	}

	static async getAllInList(listId: number): Promise<Task[]> {
		const { rows: records }: { rows: TaskRecord[] } = await DB.query(
			"SELECT * FROM tasks WHERE list_id = $1",
			[listId]
		);

		return Array.from(LinkedList(records, "id", "prev_task_id")).map(
			Task.fromRecord
		);
	}

	static fromRecord({ id, name, done, created_at, list_id }: TaskRecord): Task {
		return new Task({
			id,
			name,
			done,
			createdAt: created_at,
			listId: list_id
		});
	}

	constructor(params: TaskParams) {
		for (const key of Object.keys(params)) {
			this[key] = params[key];
		}
	}

	list(): Promise<TaskList> {
		return TaskList.findById(this.listId);
	}
}
