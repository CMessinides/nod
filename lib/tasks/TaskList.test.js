import TaskList from "./TaskList";

// For mocking
import DB from "../DB";
import Task from "./Task";
import MockRecords from "./fixtures/TaskListRecords";

jest.mock("../DB");
DB.query.mockResolvedValue({ rows: [], rowCount: 0 });

jest.mock("./Task");
Task.findById.mockResolvedValue(new Task());

describe("static methods", () => {
	describe("findById", () => {
		it("should return the list with the given ID", () => {
			const record = MockRecords[0];
			const { id } = record;

			mockDBRows([record]);

			expect(TaskList.findById(id)).resolves.toMatchSnapshot();
		});

		it("should return null when no list is found", () => {
			mockDBRows([]);

			expect(TaskList.findById(1)).resolves.toBe(null);
		});
	});
});

describe("instance methods", () => {
	describe("tasks", () => {
		it("should return the list's child tasks", () => {
			const taskList = TaskList.fromRecord(MockRecords[0]);
			const tasks = [
				new Task({ id: 1, listId: taskList.id }),
				new Task({ id: 2, listId: taskList.id })
			];

			Task.getAllInList.mockResolvedValue(tasks);

			expect(taskList.tasks()).resolves.toBe(tasks);
		});
	});
});

function mockDBRows(rows) {
	DB.query.mockResolvedValueOnce({ rows, rowCount: rows.length });
}
