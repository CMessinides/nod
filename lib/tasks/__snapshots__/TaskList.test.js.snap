// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static methods findById should return the list with the given ID 1`] = `
TaskList {
  "id": 1,
  "name": "Chores",
}
`;
